cli/tools/registry/provenance.rs

```diff
@@ -0,0 +1,622 @@
+  credentials: Credentials,
+  public_key_request: PublicKeyRequest,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct CertificateChain {
+  certificates: Vec<String>,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct SigningCertificateResponse {
+  signed_certificate_embedded_sct: Option<CertificateChain>,
+  signed_certificate_detached_sct: Option<CertificateChain>,
+}
+
+impl FulcioSigner {
+  pub fn new() -> Result<Self, AnyError> {
+    let rng = SystemRandom::new();
+    let document = EcdsaKeyPair::generate_pkcs8(ALGORITHM, &rng)?;
+    let ephemeral_signer =
+      EcdsaKeyPair::from_pkcs8(ALGORITHM, document.as_ref(), &rng)?;
+
+    Ok(Self {
+      ephemeral_signer,
+      rng,
+      client: Client::new(),
+    })
+  }
+
+  pub async fn sign(
+    self,
+    data: &[u8],
+  ) -> Result<(ring::signature::Signature, KeyMaterial), AnyError> {
+    // Request token from GitHub Actions for audience "sigstore"
+    let token = gha_request_token("sigstore").await?;
+    // Extract the subject from the token
+    let subject = extract_jwt_subject(&token)?;
+
+    // Sign the subject to create a challenge
+    let challenge =
+      self.ephemeral_signer.sign(&self.rng, subject.as_bytes())?;
+
+    let subject_public_key = self.ephemeral_signer.public_key().as_ref();
+    let algorithm = spki::AlgorithmIdentifier {
+      oid: elliptic_curve::ALGORITHM_OID,
+      parameters: Some((&p256::NistP256::OID).into()),
+    };
+    let spki = spki::SubjectPublicKeyInfoRef {
+      algorithm,
+      subject_public_key: asn1::BitStringRef::from_bytes(subject_public_key)?,
+    };
+    let pem = spki.to_pem(LineEnding::LF)?;
+
+    // Create signing certificate
+    let certificates = self
+      .create_signing_certificate(&token, pem, challenge)
+      .await?;
+
+    let signature = self.ephemeral_signer.sign(&self.rng, data)?;
+
+    Ok((
+      signature,
+      KeyMaterial {
+        _case: "x509Certificate",
+        certificate: certificates[0].clone(),
+      },
+    ))
+  }
+
+  async fn create_signing_certificate(
+    &self,
+    token: &str,
+    public_key: String,
+    challenge: ring::signature::Signature,
+  ) -> Result<Vec<String>, AnyError> {
+    let url = format!("{}/api/v2/signingCert", DEFAULT_FULCIO_URL);
+    let request_body = CreateSigningCertificateRequest {
+      credentials: Credentials {
+        oidc_identity_token: token.to_string(),
+      },
+      public_key_request: PublicKeyRequest {
+        public_key: PublicKey {
+          algorithm: "ECDSA",
+          content: public_key,
+        },
+        proof_of_possession: BASE64_STANDARD.encode(challenge.as_ref()),
+      },
+    };
+
+    let response = self.client.post(url).json(&request_body).send().await?;
+
+    let body: SigningCertificateResponse = response.json().await?;
+
+    let key = body
+      .signed_certificate_embedded_sct
+      .or_else(|| body.signed_certificate_detached_sct)
+      .unwrap();
```

Comment:
This should probably handle the error when challenge verfication fails in Fulcio.